sky/utils/kubernetes_utils.py

```diff
@@ -147,6 +159,89 @@ def get_kubernetes_nodes() -> List[Any]:
+        """Checks if the instance fits on the cluster based on CPU and memory.
+
+        We check only capacity, not allocatable, because availability can
+        change during scheduling, and we want to let the Kubernetes scheduler
+        handle that.
+        """
+        # We log max CPU and memory found on the GPU nodes for debugging.
+        max_cpu = 0.0
+        max_mem = 0.0
+
+        for node in node_list:
+            node_cpus = parse_cpu_or_gpu_resource(node.status.capacity['cpu'])
+            node_memory_gb = parse_memory_resource(
+                node.status.capacity['memory'], unit='G')
+            if node_cpus > max_cpu:
+                max_cpu = node_cpus
+                max_mem = node_memory_gb
+            if node_cpus >= candidate_instance_type.cpus and \
+                    node_memory_gb >= candidate_instance_type.memory:
+                return True, None
+        return False, (
+            'Maximum resources found on a single node: '
+            f'{max_cpu} CPUs, {common_utils.format_float(max_mem)}G Memory')
+
+    nodes = get_kubernetes_nodes()
+    k8s_instance_type = kubernetes_cloud.KubernetesInstanceType.\
+        from_instance_type(instance)
+    acc_type = k8s_instance_type.accelerator_type
+    if acc_type is not None:
+        # If GPUs are requested, check if GPU type is available, and if so,
+        # check if CPU and memory requirements on the specific node are met.
+        try:
+            gpu_label_key, gpu_label_val = get_gpu_label_key_value(acc_type)
+        except exceptions.ResourcesUnavailableError as e:
+            # If GPU not found, return empty list and error message.
+            return False, str(e)
+        # Get the set of nodes that have the GPU type
+        gpu_nodes = [
+            node for node in nodes if gpu_label_key in node.metadata.labels and
+            node.metadata.labels[gpu_label_key] == gpu_label_val
+        ]
+        assert len(gpu_nodes) > 0, 'GPU nodes not found'
+        # Check if the CPU and memory requirements are met on at least one node
+        fits, reason = check_cpu_mem_fits(k8s_instance_type, gpu_nodes)
+        if not fits:
+            if reason is not None:
+                reason_prefix = (f'GPU nodes with {acc_type} do not have '
+                                 'enough CPU and/or memory. ')
+                reason = reason_prefix + reason
+            return fits, reason
+        else:
+            return fits, reason
+    else:
+        # If no GPU is requested, check if CPU and memory requirements are met
+        # on at least one node.
+        fits, reason = check_cpu_mem_fits(k8s_instance_type, nodes)
+        if not fits:
+            if reason is not None:
+                reason_prefix = 'No nodes found with enough CPU and/or memory. '
+                reason = reason_prefix + reason
+            return fits, reason
+        else:
+            return fits, reason
```

Comment:
nit: Seems most of this part are similar, should we combine them?

Reply:
Ah good point - I've cleaned this up